File: small_problems/easy_9/8.rb
=begin
Input: array (of nested arrays)
Output: array of strings (correct quantity of each fruit)


Rules:
- each nested array is 2-elements: fruit, quantity
- return value should be an array of separate fruit strings, each fruti listed the quantity number of times from the input

ALGORITHM:
- given an array, "fruits"
- create an empty array, "grocery_list"
- iterate through fruits array
  - add the current fruit to grocery_list the given number of times
- return grocery list array
=end

# First Solution - #each
def buy_fruit(fruits)
  grocery_list = []
  fruits.each do |fruit, count|
    count.times { grocery_list << fruit }
  end
  grocery_list
end

# Second Solution - #each_with_object
def buy_fruit(fruits)
  fruits.each_with_object([]) do |(fruit, count), list|
    count.times { list << fruit }
  end
end

# LS Solution - #map
def buy_fruit(list)
  list.map { |fruit, quantity| [fruit] * quantity }.flatten
end

p buy_fruit([["apples", 3], ["orange", 1], ["bananas", 2]]) ==
  ["apples", "apples", "apples", "orange", "bananas","bananas"]

File: js_small_problems/list-processing/8.js
/*
Grocery List

Write a function that takes a grocery list in a two-dimensional array and returns a one-dimensional array. Each element in the grocery list contains a fruit name and a number that represents the desired quantity of that fruit. The output array is such that each fruit name appears the number of times equal to its desired quantity.

In the example below, we want to buy 3 apples, 1 orange, and 2 bananas. Thus, we return an array that contains 3 apples, 1 orange, and 2 bananas.

INPUT: nested array
OUTPUT: 1D array
  - fruit name appears n times, where n is the second element of each nested array

Abstractions:
- transform the array elements into the expanded grocery list
- flatten the array of arrays using a folding strategy
*/

function repeat(fruit) {
  let result = [];

  for (let count = 0; count < fruit[1]; count += 1) {
    result.push(fruit[0]);
  }

  return result;
}

function buyFruit(fruits) {
  return fruits.reduce((result, fruit) => {
    return result.concat(repeat(fruit));
  }, []);
}

// LS Solution
// function buyFruit(fruitsList) {
//   return fruitsList.map(fruit => repeat(fruit))
//                    .reduce((groceryList, fruit) => groceryList.concat(fruit));
// }

// function repeat(fruit) {
//   const result = [];

//   for (let i = 0; i < fruit[1]; i += 1) {
//     result.push(fruit[0]);
//   }

//   return result;
// }

console.log(buyFruit([['apple', 3], ['orange', 1], ['banana', 2]]));
// returns ["apple", "apple", "apple", "orange", "banana", "banana"]
